### PR TITLE
Comprehensive F extension support with SoftFloat integration

### LINKED ISSUE #78: any idea why instruction like `fadd` is not passing some of the test cases such as `fadd_b10-01` ?
is there some corner cases need to be considered?

### PR DESCRIPTION
**Pull request by visitorckw**
This PR aimed at achieving full F extension support within rv32emu. It incorporates the SoftFloat open-source library into the rv32emu project to provide comprehensive support for the F extension. 

The key changes in this PR include:
1. Integrated the SoftFloat library into rv32emu, completing the F extension support within the project.
2. Introduced F extension testing in the CI pipeline, ensuring thorough validation of the implemented functionality.
3. Updated the README to correct the description of F extension support from incomplete to full support.

Close: #78

### CODE REVIEW

**Comment by jserv** [comment]
Is it possible to use git submodule instead of copying SoftFloat files in this repository?

**Comment by jserv** [comment]
> The key changes in this PR include:
> 1. Integrated the SoftFloat library into rv32emu, completing the F extension support within the project.

I would like to revoke my opinion on the implementation of the RISC-V F extension. Integrating SoftFloat would achieve full compatibility and reduce maintenance efforts. There is no need to retain the original RV32F-specific code. That means the conditional build between the original and SoftFloat-specific code can be removed, and let's focus on SoftFloat.

**Comment by visitorckw** [comment]
> Is it possible to use git submodule instead of copying SoftFloat files in this repository?

Currently, I've copied all files as I added additional files in the SoftFloat folder to comply with the risc-v F extension standard. Perhaps we can explore alternatives, such as copying these files into the SoftFloat directory during compilation.

**Comment by visitorckw** [comment]
> I would like to revoke my opinion on the implementation of the RISC-V F extension. Integrating SoftFloat would achieve full compatibility and reduce maintenance efforts. There is no need to retain the original RV32F-specific code. That means the conditional build between the original and SoftFloat-specific code can be removed, and let's focus on SoftFloat.

Sounds like a good idea, I'll make the change later.

**Comment by jserv** [comment]
> Currently, I've copied all files as I added additional files in the SoftFloat folder to comply with the risc-v F extension standard. Perhaps we can explore alternatives, such as copying these files into the SoftFloat directory during compilation.

To incorporate external source code, please refer to the `Makefile` and examine the sections related to `mini-gdbstub`. These sections contain information on how to utilize `git submodule` and their associated build rules.

File: Makefile [hunk 1]
```diff
@@ -211,5 +223,6 @@ distclean: clean
 	$(RM) *.zip
 	$(RM) -r $(OUT)/mini-gdbstub
 	-$(RM) $(OUT)/.config
+	$(MAKE) -C src/SoftFloat-3e/build/Linux-RISCV-GCC/ clean
```
Comment (jserv) [inline Makefile]: It should begin with a `-` character, indicating that it is acceptable to fail.
See https://www.gnu.org/software/make/manual/html_node/Errors.html

File: .gitmodules [hunk 2]
```diff
@@ -4,3 +4,6 @@
 [submodule "mini-gdbstub"]
 	path = src/mini-gdbstub
 	url = https://github.com/RinHizakura/mini-gdbstub
+[submodule "SoftFloat-3e"]
+	path = src/SoftFloat-3e
```
Comment (jserv) [inline .gitmodules]: Simply name it `src/softfloat` as QEMU does.
See https://github.com/qemu/qemu/blob/master/fpu/softfloat.c

File: src/riscv.h [hunk 3]
```diff
@@ -6,6 +6,7 @@
 #pragma once
 #include <stdbool.h>
 #include <stdint.h>
+#include "SoftFloat-3e/source/include/softfloat.h"
```
Comment (jserv) [inline src/riscv.h]: The header inclusion is a bit lenthy. Can you shorten it?

**Comment by jserv** [comment]
Can you review `src/softfloat.h` and determine whether we should retain it (partially) or not?

**Comment by visitorckw** [comment]
> Can you review `src/softfloat.h` and determine whether we should retain it (partially) or not?

We can remove the is_snan function as the softfloat library already provides the same functionality with isSignalingNaN. However, the rest of the content needs to be retained. Additionally, I'm considering whether renaming src/softfloat.h could prevent confusion.

**Comment by visitorckw** [comment]
I'm uncertain why the current code is failing the Codacy Security Scan test.

File: .gitmodules [hunk 4]
```diff
@@ -4,3 +4,6 @@
 [submodule "mini-gdbstub"]
 	path = src/mini-gdbstub
 	url = https://github.com/RinHizakura/mini-gdbstub
+[submodule "softfloat"]
+	path = src/softfloat
+	url = https://github.com/visitorckw/SoftFloat-3e.git
```
Comment (jserv) [inline .gitmodules]: Use shorter URL. i.e., https://github.com/visitorckw/SoftFloat-3e`
Can you utilize https://github.com/ucb-bar/berkeley-softfloat-3 instead?

File: src/softfloat.h [hunk 5]
```diff
@@ -85,9 +85,32 @@ static inline bool is_nan(uint32_t f)
     return (expn == FMASK_EXPN && frac);
 }
 
-static inline bool is_snan(uint32_t f)
-{
-    const uint32_t expn = f & FMASK_EXPN;
-    const uint32_t frac = f & FMASK_FRAC;
-    return (expn == FMASK_EXPN && frac && !(frac & FMASK_QNAN));
-}
+#define set_fflag()                                              \
```
Comment (jserv) [inline src/softfloat.h]: Can you rewrite it by using a `static inline` function?

File: .gitmodules (hunk 4 above)
Comment (visitorckw) [reply .gitmodules]: The Makefile for building softfloat.a at https://github.com/ucb-bar/berkeley-softfloat-3 doesn't specify an output path parameter. As a result, using it may require keeping softfloat.a within the src/softfloat instead of the build/softfloat. Additionally, the previously mentioned issue of lengthy header inclusion in riscv.h might be temporarily challenging to resolve.

**Comment by jserv** [comment]
> I'm uncertain why the current code is failing the Codacy Security Scan test.

It seems that something is wrong with Docker Hub.

File: .gitmodules (hunk 4 above)
Comment (jserv) [reply .gitmodules]: Rather than duplicating files without their historical commits, consider forking `ucb-bar/berkeley-softfloat-3` and incorporating your own modifications and bug fixes. Does this approach seem reasonable?

Comment (visitorckw) [reply .gitmodules]: Agreed. I'll make the adjustment.

**Comment by jserv** [comment]
When rv32emu is built without F extension, i.e., `make ENABLE_EXT_F=0`, I got the following error messages:
```
src/riscv.h:9:10: fatal error: softfloat/softfloat.h: No such file or directory
    9 | #include "softfloat/softfloat.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Review by jserv — CHANGES_REQUESTED** [review]
Fix build without F extension support.

**Comment by visitorckw** [comment]
Apologies for the error. I've corrected it. It should be resolved now.

**Comment by jserv** [comment]
Thank @visitorckw for contributing!
